axelor-stock/src/main/java/com/axelor/apps/stock/db/repo/StockLocationStockRepository.java

```diff
@@ -64,32 +64,35 @@ public StockLocation save(StockLocation stockLocation) {
     // Barcode generation
     AppStock appStock = appStockService.getAppStock();
 
-    if (appStock != null
-        && appStock.getActivateStockLocationBarCodeGeneration()
-        && stockLocation.getBarCode() == null
-        && stockLocation.getBarcodeTypeConfig() != null
-        && stockLocation.getSerialNumber() != null) {
-      try {
-        boolean addPadding = false;
-        BarcodeTypeConfig barcodeTypeConfig = stockLocation.getBarcodeTypeConfig();
-        if (appStock.getEditStockLocationBarcodeType() == false) {
-          barcodeTypeConfig = appStock.getStockLocationBarcodeTypeConfig();
-        }
-        InputStream inStream =
-            barcodeGeneratorService.createBarCode(
-                stockLocation.getSerialNumber(), barcodeTypeConfig, addPadding);
-        if (inStream != null) {
-          MetaFile barcodeFile =
-              metaFiles.upload(
-                  inStream, String.format("StockLocationBarCode%d.png", stockLocation.getId()));
-          stockLocation.setBarCode(barcodeFile);
+    if (appStock != null && appStock.getActivateStockLocationBarCodeGeneration()) {
+
+      BarcodeTypeConfig barcodeTypeConfig = stockLocation.getBarcodeTypeConfig();
+      // Apply defaulting if needed
+      if (appStock.getEditStockLocationBarcodeType() == false) {
+        barcodeTypeConfig = appStock.getStockLocationBarcodeTypeConfig();
+      }
+
+      if (stockLocation.getBarCode() == null
+          && barcodeTypeConfig != null
+          && stockLocation.getSerialNumber() != null) {
+        try {
+          boolean addPadding = false;
+          InputStream inStream =
+              barcodeGeneratorService.createBarCode(
+                  stockLocation.getSerialNumber(), barcodeTypeConfig, addPadding);
+          if (inStream != null) {
+            MetaFile barcodeFile =
+                metaFiles.upload(
+                    inStream, String.format("StockLocationBarCode%d.png", stockLocation.getId()));
+            stockLocation.setBarCode(barcodeFile);
+          }
```

Comment:
This block of code is duplicated in ProductBaseRepository, OperationOrderService and TrackingNumberManagementRepository, so please create a new method in barcodeGeneratorService to create and upload the barcode file

Reply:
Done. I have implemented a new method on BarcodeGeneratorService and have updated all impacted code.